README.md
# Bar - VS Code

![Downloads](https://img.shields.io/vscode-marketplace/d/olback.bar.svg)

## Features
The goal with this extension is to make it easier to develop binary files. In the status bar you have the option to Build, Run and "Build and run" the project.

![status bar buttons](https://raw.githubusercontent.com/olback/bar-vscode/master/images/status_bar.png)  

## Extension Commands
* `Bar: Init`: Initialize the extension.
* `Bar: Reload`: Alias for `Bar: Init`.
* `Bar: Build`: Build your project.
* `Bar: Run`: Run your project.
* `Bar: Build and run project`: Build and run project.
* `Bar: Reset` Reset the config file.
* `Bar: Edit config` Open the config file.

## Extension Keybindings
* Build and run: `shift+f6`
* Build: `ctrl+shift+f6`
* Run: `shift+f2`

## Configuration
Open the command pallete (Cmd or Ctrl+Shift+P) and run `Bar: Init` to initialize Bar and create the config file.

The commands for building/compiling and running your project are saved in `projectRoot/.vscode/bar.conf.json`.

```
{
    "commands": {
        "build": "your-build-command",
        "run": "./project-exe"
    },
    "messages": {
        "building": true,
        "buildSuccess": true,
        "buildError": true,            
        "buildErrorMessageBox": true, 
        "run": true
    }
}
```

The values in the "messages" object allow you to control the kinds of status messages that will be shown. When `buildErrorMessageBox` is true, a small informational error box will also appear in the bottom right corner when the build fails.

## Bugs
Please report any bugs or issues on [GitHub](https://github.com/olback/bar-vscode)

**Enjoy!**


CHANGELOG.md
# Change Log

#  0.2.0
* Thanks to @nathan815 for cleaning up my mess.
* Added ability to disable messages.

## 0.1.1
* Fixed config-write bug

## 0.1.0
* Use `vscode.workspace.worspaceFolders[0].uri.fsPath` instead of `vscode.workspace.rootPath`.

## 0.0.9
* Send CTRL-C to terminal before sending command. Thanks @agurz.

## 0.0.8
* Clear output before running the build command.

## 0.0.7
* Cleaned up some of the code.
* Added icons to the status bar items.
* 'Build and run' is now shortend to 'Bar'.

## 0.0.6
* Fixed 'Add new config?' bug.

## 0.0.5
* Run commands are now run from the integrated terminal. Output will also be shown here.
* Added `Bar: Reload`. Alias for `Bar: Init`.

## 0.0.4
* Added keyboard shortcut to build and run: `shift+f6`.
* Added keyboard shortcut to build: `ctrl+shift+f6`.
* Added keyboard shortcut to run: `shift+f2`.
* Added `Bar: Edit config` command.

## 0.0.3
* When the build fails, show the output in Bar Output.
* Ask before creating bar.conf.json
* Auto load the extension if `.vscode/bar.conf.json` exists.
* Can't initialize the extension more than once. (bug fix)
* Minor improvements.
* Added MIT License.
* Added more info in `package.json`.
* Added more bugs.

## 0.0.2
* Updated README.md.
* Added `Bar: Reset` command.

## 0.0.1
* Initial release of Bar!


src/extension.ts
'use strict';

import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';

const ex = 'bar';

let output: vscode.OutputChannel;
let terminal: vscode.Terminal;
let config: any;
const statusBarItems: Array<vscode.StatusBarItem> = [];

function load(): void {
    // Do all the setup stuff

    if (vscode.workspace.workspaceFolders === undefined) {
        return;
    }

    // config = JSON.parse(fs.readFileSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json'), 'utf8'));
    try {

        config = JSON.parse(fs.readFileSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json'), 'utf8'));

    } catch (e) {

        vscode.window.showErrorMessage(`Bar config reload failed: ${e}`);

    }

    output = vscode.window.createOutputChannel('Bar');
    terminal = vscode.window.createTerminal('Bar');

    addStatusBarItem("$(package)  Build", "bar.build", "Build project");
    addStatusBarItem("$(terminal)  Run", "bar.run", "Run project/file");
    addStatusBarItem("$(rocket)  Bar", "bar.bar", "Build and run project/file");
}

function addStatusBarItem(str: string, cmd?: string, tip?: string, col?: string) { // (name, command, tooltip, color)
    statusBarItems.push(vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left));
    statusBarItems[statusBarItems.length-1].text = str;
    if (cmd) {
        statusBarItems[statusBarItems.length-1].command = cmd;
    }
    if (tip) {
        statusBarItems[statusBarItems.length-1].tooltip = tip;
    }
    if (col) {
        statusBarItems[statusBarItems.length-1].color = col;
    }
    statusBarItems[statusBarItems.length-1].show();
}

export function activate(context: vscode.ExtensionContext): void {

    context.subscriptions.push(
        vscode.commands.registerCommand(`${ex}.init`, () => {

            if (vscode.workspace.workspaceFolders === undefined) {
                return vscode.window.showErrorMessage('Bar requires a workspace.');
            }

            if (!fs.existsSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode'))) {
                fs.mkdirSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode'));
            }

            if (!fs.existsSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json'))) {
                fs.writeFileSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json'), JSON.stringify({
                    commands: {
                        build: 'echo "build command"',
                        run: 'echo "run command"',
                    }
                }, null, 4) + '\n', 'utf8');
            }

            vscode.window.showTextDocument(vscode.Uri.file(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json')));

            if (statusBarItems.length === 0) {
                load();
            }

        }),
        vscode.commands.registerCommand(`${ex}.build`, () => {
            output.appendLine(`[Info] Building with command ${config.commands.build}`);
            terminal.show();
            terminal.sendText('\x03'); // https://stackoverflow.com/questions/5774689/what-is-003-special-for
            terminal.sendText(config.commands.build);
        }),
        vscode.commands.registerCommand(`${ex}.run`, () => {
            output.appendLine(`[Info] Running with command ${config.commands.run}`);
            terminal.show();
            terminal.sendText('\x03');
            terminal.sendText(config.commands.run);
        }),
        vscode.commands.registerCommand(`${ex}.bar`, () => {
            console.log('wip');
        }),
    );

    if (vscode.workspace.workspaceFolders && fs.existsSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json'))) {

        load();

        // const w = fs.watch(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json'), 'utf8');
        // w.addListener('change', () => {

        //     if (vscode.workspace.workspaceFolders) {

        //         try {

        //             config = JSON.parse(fs.readFileSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json'), 'utf8'));
        //             vscode.window.showInformationMessage('<<<<Bar config reloaded');

        //         } catch (e) {

        //             vscode.window.showErrorMessage(`<<<<Bar config reload failed: ${e}`);

        //         }

        //     }

        // });

    }

    console.log('bar is now active!');

}

// this method is called when your extension is deactivated
export function deactivate() {
}


extension.js
/**
 *  © olback 2017
 *      BAR
 */

// Imports
const vscode = require('vscode');
const exec = require('child_process').exec;
const path = require('path');
const fs = require('fs');

const DEFAULT_CONFIG = {
    commands: {
        build: "",
        run: ""
    },
    messages: {
        building: true,
        buildSuccess: true,
        buildError: true,
        buildErrorMessageBox: true,
        run: true
    }
};

const STATUS_MESSAGE_DURATION = 2000;

let config = DEFAULT_CONFIG;
let statusbar = 0;
let resetConfigVal = 0;
let newBuildCommand;
let newRunCommand;
const statusBarItems = {};
const configPath = path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode', 'bar.conf.json');
const output = vscode.window.createOutputChannel('Bar');
const terminal = vscode.window.createTerminal('Bar');

function addStatusBarItem(key, name, cmd, tip, col) { // (key, name, command, tooltip, color)
    const item = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
    if(name) item.text = name;
    if(cmd) item.command = cmd;
    if(tip) item.tooltip = tip;
    if(col) item.color = col;
    item.show();
    statusBarItems[key] = item;
    return item;
}

let statusMessageTimeout;
function showStatusMessage(str, duration = STATUS_MESSAGE_DURATION) {
    clearTimeout(statusMessageTimeout);
    statusBarItems.statusMessage.text = str;
    statusBarItems.statusMessage.show();
    if(duration) {
        statusMessageTimeout = setTimeout(() => {
            statusBarItems.statusMessage.hide();
        }, duration);
    }
}

function addStatusBar() {
    addStatusBarItem("divider1", "|");
    addStatusBarItem("build", "$(package)  Build", "bar.build", "Build project");
    addStatusBarItem("run", "$(terminal)  Run", "bar.run", "Run project/file");
    addStatusBarItem("bar", "$(rocket)  Bar", "bar.bar", "Build and run project/file");
    addStatusBarItem("divider2", "|");
    addStatusBarItem("statusMessage");
}

function newConfigBuild() {
    if(resetConfigVal == 0) {
        vscode.window.showInformationMessage('Add Bar config?', 'Yes')
        .then(selection => {
            if(selection == "Yes") {
                vscode.window.showInputBox({prompt: 'Command to build the project. Example: "make build"', ignoreFocusOut: true})
                .then(val => { newBuildCommand = val; newConfigRun(); });
            }
        });
    } else {
        vscode.window.showInputBox({prompt: 'Command to build the project. Example: "make build"', ignoreFocusOut: true})
        .then(val => { newBuildCommand = val; newConfigRun(); });
    }
}

function newConfigRun() {
    vscode.window.showInputBox({prompt: "Command to run your project. Example: \"./your-executable-file\"", ignoreFocusOut: true})
        .then(val => { newRunCommand = val; writeConfig(); });
}

function writeConfig() {
    console.log('Creating new config: ', configPath);
    if (!fs.existsSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode'))) {
        fs.mkdirSync(path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, '.vscode'))
    }

    const newConfig = { ...DEFAULT_CONFIG };
    newConfig.commands.build = newBuildCommand;
    newConfig.commands.run = newRunCommand;

    fs.writeFileSync(configPath, JSON.stringify(newConfig, null, 4));
    readConfig();
    vscode.window.showInformationMessage('Bar done! Saved settings to ' + configPath);
}

function readConfig() {
    if(fs.existsSync(configPath)) {
        const userConfig = JSON.parse(fs.readFileSync(configPath, 'utf8'));
        config = Object.assign({}, DEFAULT_CONFIG, userConfig);
        if(statusbar == 0) {
            addStatusBar();
            statusbar = 1;
        }
    } else {
        newConfigBuild();
    }
}

function resetConfig() {
    vscode.window.showWarningMessage('Reset Bar config?', 'Yes').then(selection => {
      if(selection == "Yes") {
        if(fs.existsSync(configPath)) {
            fs.unlinkSync(configPath, (err) => {
                if(err) return console.log(err);
                console.log("Removed " + configPath);
            });
            resetConfigVal = 1;
            init();
        }
    }
    });
}

function init() {
    readConfig();
}

function build() {
    return new Promise((resolve) => {
        output.clear();
        if(config.messages.building) {
            showStatusMessage('Building...', 0);
        }

        let ls = exec(config.commands.build, {
            cwd: vscode.workspace.workspaceFolders[0].uri.fsPath, 
            maxBuffer: 2048000
        });

        ls.on('close', (code) => {
            //console.log(`child process exited with code ${code}`);
            statusBarItems.statusMessage.hide();
            if(code == 0) {
                resolve();
                if(config.messages.buildSuccess) {
                    showStatusMessage('Build Successful');
                }
            } else {
                if(config.messages.buildError) {
                    showStatusMessage('Build Failed');
                }
                if(config.messages.buildErrorMessageBox) {
                    vscode.window.showErrorMessage('Build Failed. Check Bar Output.');
                }
            }
        });

        ls.stderr.on('data', (data) => {
            //console.log(`stderr: ${data}`);
            output.show();
            output.appendLine(data);
        });
    });
}

function run() {
    if(config.messages.run) {
        showStatusMessage('Running');
    }
    //exec(config.commands.run, {cwd: vscode.workspace.workFolders[0].uri.fsPath, maxBuffer: 2048000});
    terminal.show();
    terminal.sendText('\003'); // Ctrl+C https://stackoverflow.com/questions/5774689/what-is-003-special-for
    terminal.sendText(config.commands.run);
}

function editConfig() {
    let openPath = vscode.Uri.file(configPath);
    vscode.workspace.openTextDocument(openPath).then(doc => {
      vscode.window.showTextDocument(doc);
    });
}

// this method is called when your extension is executed
function activate(context) {

    console.log('bar is active!');

    // Build and run
    let disposable = vscode.commands.registerCommand('bar.bar', () => {
        build().then(run);
    });
    context.subscriptions.push(disposable);

    context.subscriptions.push(
        vscode.commands.registerCommand('bar.init', init),
        vscode.commands.registerCommand('bar.reload', init),
        vscode.commands.registerCommand('bar.build', build),
        vscode.commands.registerCommand('bar.run', run),
        vscode.commands.registerCommand('bar.reset', resetConfig),
        vscode.commands.registerCommand('bar.config', editConfig)
    );

    readConfig();

}

// this method is called when your extension is deactivated
function deactivate() {
    //console.log('bar unloaded');
}

exports.activate = activate;
exports.deactivate = deactivate;
